containers-tests/tests/Utils/IsUnit.hs

```diff
@@ -0,0 +1,16 @@
+{-# LANGUAGE MagicHash #-}
+
+module Utils.IsUnit (isUnit) where
+
+import GHC.Exts
+
+same :: a -> a -> Bool
+same x y = case reallyUnsafePtrEquality# x y of
+    0# -> False
+    _  -> True
+
+-- | Check whether the argument is a fully evaluated unit `()`.
+--
+-- Uses `reallyUnsafePtrEquality#`.
+isUnit :: () -> Bool
+isUnit = same ()
```

Comment:
I'm wondering whether I'm reinventing a wheel here. This seems closely related to what ChasingBottoms is doing, but I didn't find anything like this there.

Reply:
This smells funny/fragile to me. I would generally feel more comfortable with an approach that uses `unsafePerformIO` to find out when/how many times things are evaluated.

Reply:
Oh, but of course we can't always do that .... Hrmmm. Could you at least rename `same` to match our internal pointer equality function?

Reply:
Oh I wasn't aware of `ptrEq`, or I would have used that; will change.

I should probably make the tests conditional on __GLASGOW_HASKELL__ being defined... and  maybe make isUnit tristate? (Yes | No | Maybe; the last indicating that the result is unknown.)

Reply:
> Oh I wasn't aware of `ptrEq`, or I would have used that; will change.

Ah it's not so easy because the module is hidden, and we probably want to keep the library parts of `containers-tests` and `containers` the same.

> I should probably make the tests conditional on `__GLASGOW_HASKELL__` being defined... and maybe make isUnit tristate? (Yes | No | Maybe; the last indicating that the result is unknown.)

I will go with `Maybe Bool`.

Reply:
Given your later comment, I stuck with `Bool`. I did add some logic to deal with non-ghc implementations though.